python/python/tests/test_pubsub.py

```diff
@@ -0,0 +1,486 @@
+                            for message, cnt in unsharded_channel_messages.items():
+                                if message not in self.messages_to_receive[channel_and_pattern]:
+                                    self.messages_to_receive[channel_and_pattern][message] = 0
+                                self.messages_to_receive[channel_and_pattern][message] += cnt
+
+        # TODO: Allow async fanning out?
+        def sched_publish_messages(self) -> List[Awaitable[None]]:
+            # coroutines = []
+            # for sharded_channel, sharded_messages_to_publish in self.sharded_messages_to_publish:
+            #     for message in sharded_messages_to_publish:
+            #         coroutines += self.redis_client.publish(message=message, channels={sharded_channel}, sharded=True)
+
+            # for unsharded_channel, unsharded_messages_to_publish in self.unsharded_messages_to_publish:
+            #     for message in unsharded_messages_to_publish:
+            #         coroutines += self.redis_client.publish(message=message, channels={unsharded_channel}, sharded=False)
+
+            # return coroutines
+            pass
+        
+        async def publish_messages(self) -> List[Awaitable[None]]:
+            for sharded_channel, sharded_messages_to_publish in self.sharded_messages_to_publish:
+                for message, cnt in sharded_messages_to_publish:
+                    for i in range(cnt):
+                        await self.redis_client.publish(message=message, channels={sharded_channel}, sharded=True)
+
+            for unsharded_channel, unsharded_messages_to_publish in self.unsharded_messages_to_publish:
+                for message, cnt in unsharded_messages_to_publish:
+                    for i in range(cnt):
+                        await self.redis_client.publish(message=message, channels={unsharded_channel}, sharded=False)
+
+
+    @staticmethod
+    async def publish_all_messages(all_clients: List[TestPubSub.PubSubClient]):
+        coroutines = []
+        for client in all_clients:
+            coroutines += client.sched_publish_messages()
+        asyncio.wait(coroutines)
+        
+
+    @staticmethod
+    async def publish_and_assert(all_clients: List[TestPubSub.PubSubClient]):
+        for client in all_clients:
+            client.publish_messages()
+
+        #TODO - Wait for how long?
+        for client in all_clients:
+            client.calculate_expected_messages(all_clients=all_clients)
+            client.assert_client()
+
+
+    @pytest.mark.parametrize("cluster_mode", [True, False])
+    @pytest.mark.parametrize("sharded_pubsub", [True, False])
+    async def test_pubsub_exact(self, request, cluster_mode, sharded_pubsub):
+        """ PUBSUB basic happy case using exact channel names - tests that clients receive all the messages using exact subscribe to a single channel """
+
+        CHANNEL_NAME = "test-channel"
+        CLIENTS_COUNT = 5
```

Comment:
the simple tests should have only a single client - if they fail, we want debugging to also be simple.

Reply:
The easiest functional setting is 2, will change to that

Reply:
why is 2 easier than 1?

Reply:
of course it is not easier,  it is more functional

Reply:
what does "more functional" mean here?
it does more? I agree, but a test should do as little as possible, in order to focus the coder on the source of the error if the test fails. In that sense, adding also gets, sets, and other calls to this test would make it more functional, but it doesn't help the test be focused.

Reply:
I do not agree that the test has to do as little as possible. The value of such tests are pretty low.

Reply:
I also think we shall have a basic and simple test for each option (sharded/pattern/exact), a simple subscribe publish unsbuscribe. We shall have more complicated tests as well, but having a straightforward simple test will give us better visibility with failures

Reply:
The value of simple tests is that if they fail, it's easy to understand why they fail. The value of complex tests is that they can catch multiple points of failure. The value of a combination of complex & simple tests is that if the simple test failed, you can start by fixing it, and that will save you time debugging a more complex test - but if the simple test didn't fail, it still saves you time debugging the complex test, because it removes at least one scenario that needs to be considered. If you have multiple simple tests, this value aggregates.

So, we're not saying that you should have only simple tests, but please - ALSO add simple tests.

Reply:
added based test test_pubsub_basic()

Reply:
done